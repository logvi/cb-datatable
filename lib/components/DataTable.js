'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _from = require('babel-runtime/core-js/array/from');

var _from2 = _interopRequireDefault(_from);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _utils = require('../utils');

var _ = _interopRequireWildcard(_utils);

var _TableBody = require('./TableBody');

var _TableBody2 = _interopRequireDefault(_TableBody);

var _TableHeader = require('./TableHeader');

var _TableHeader2 = _interopRequireDefault(_TableHeader);

var _Pager = require('./Pager');

var _Pager2 = _interopRequireDefault(_Pager);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// type Props = {
//     data: DataT,
//     columns: Array<ColumnT>,
//     rowsPerPage: number,
//     orderBy: {[key: string]: string},
// 	className: string,
// 	fixedHeader: bool,
// 	displayAllData: bool,
// 	onClick: any,
// 	onRowRender: any
// };


// $FlowSkipCheck
var DataTable = function (_Component) {
    (0, _inherits3.default)(DataTable, _Component);

    function DataTable(props) {
        (0, _classCallCheck3.default)(this, DataTable);

        var _this = (0, _possibleConstructorReturn3.default)(this, (DataTable.__proto__ || (0, _getPrototypeOf2.default)(DataTable)).call(this, props));

        _this.containerScrollHandler = function (e) {
            console.log(e);
        };

        _this.selectPageHandler = function (e) {
            e.preventDefault();
            var toPage = e.target.value;
            _this.setState({ currentPage: +toPage });
        };

        _this.onHeaderAction = function (type, params) {
            switch (type) {
                case 'sort':
                    var newSorting = (0, _defineProperty3.default)({}, params.column.name, params.orderBy);
                    _this.setState({ orderBy: newSorting });
                    break;
                default:
                    break;
            }
        };

        _this.bodyScrollHandler = function (_ref) {
            var target = _ref.target;
            var scrollLeft = target.scrollLeft;

            if (_this.props.fixedHeader) {
                _this.tableHeaderComponent.table.style.transform = 'translate3d(' + -1 * scrollLeft + 'px, 0, 0)';
            }
            if (_this.props.displayAllData) {
                var scrollTop = target.scrollTop,
                    offsetHeight = target.offsetHeight,
                    scrollHeight = target.scrollHeight;

                var scrollPosToUpdate = scrollHeight * 0.15;

                if (scrollHeight - (scrollTop + offsetHeight) <= scrollPosToUpdate) {
                    _this.setPage(_this.state.currentPage + 1);
                }
            }
        };

        _this.adjustColumnsSize = function () {
            if (!_this.props.fixedHeader) return;
            // this.tableHeaderComponent.table.style.width = this.tableHeaderComponent.fixedContainer.offsetWidth + 'px';
            var headerColumns = _this.tableHeaderComponent.row.childNodes;
            // debugger;
            for (var j = 0; j < headerColumns.length; j++) {
                var bodyColumn = _this.tableBodyComponent.table.childNodes[0].childNodes[j],
                    headerColumn = headerColumns[j];
                // clear minWidth
                if (!!bodyColumn.style.width) {
                    bodyColumn.style.width = '';
                    bodyColumn.style.minWidth = '';
                    headerColumn.style.width = '';
                }
                var bodyColumnWidth = bodyColumn.getBoundingClientRect().width,
                    headerColumnWidth = headerColumn.getBoundingClientRect().width,
                    newWidth = bodyColumnWidth;
                // newWidth = Math.max(bodyColumnWidth, headerColumnWidth);
                headerColumn.style.width = newWidth + 'px';
                if (headerColumn.getBoundingClientRect().width > newWidth) {
                    newWidth = Math.max(bodyColumnWidth, headerColumnWidth);
                    headerColumn.style.width = newWidth + 'px';
                }
                bodyColumn.style.minWidth = newWidth + 'px';
                bodyColumn.style.width = newWidth + 'px';
            }
            // this.tableHeaderComponent.table.style.width = 'auto';
        };

        _this.previousClickHandler = function () {
            _this.setPage(_this.state.currentPage - 1);
        };

        _this.nextClickHandler = function () {
            _this.setPage(_this.state.currentPage + 1);
        };

        _this.state = {
            currentPage: 1,
            orderBy: props.orderBy,
            firstIndex: props.firstIndex
        };

        _this.setTableData({
            page: _this.state.currentPage,
            orderBy: _this.state.orderBy
        });
        _this.initPagesArray();
        return _this;
    }

    (0, _createClass3.default)(DataTable, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            if (this.props.fixedHeader) this.adjustColumnsSize();

            this.container.addEventListener('scroll', this.containerScrollHandler);
        }
    }, {
        key: 'componentWillUpdate',
        value: function componentWillUpdate(nextProps, nextState) {
            // console.time('datatable');
            var pageChanged = this.state.currentPage !== nextState.currentPage,
                sortingChanged = this.state.orderBy !== nextState.orderBy;

            if (pageChanged || sortingChanged) {
                this.setTableData({
                    data: nextProps.data,
                    page: pageChanged ? nextState.currentPage : null,
                    orderBy: sortingChanged ? nextState.orderBy : this.state.orderBy
                });
                return;
            }

            this.setTableData({ data: nextProps.data, orderBy: this.state.orderBy });
            this.initPagesArray(nextProps.data);
        }
    }, {
        key: 'componentDidUpdate',
        value: function componentDidUpdate() {
            if (this.props.fixedHeader) {
                // debugger;
                this.adjustColumnsSize();
            }
        }
    }, {
        key: 'getPagesCount',
        value: function getPagesCount(data) {
            var _data = data || this.props.data;
            return Math.ceil(_data.length / this.props.rowsPerPage);
        }
    }, {
        key: 'initPagesArray',
        value: function initPagesArray(data) {
            var _data = data || this.props.data;
            this.pages = (0, _from2.default)(new Array(this.getPagesCount(_data)), function (val, index) {
                return index + 1;
            });
        }
    }, {
        key: 'setTableDataByFirstIndex',
        value: function setTableDataByFirstIndex() {}
    }, {
        key: 'setTableData',
        value: function setTableData() {
            var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
                data = _ref2.data,
                page = _ref2.page,
                orderBy = _ref2.orderBy,
                rowsPerPage = _ref2.rowsPerPage,
                _ref2$displayAllData = _ref2.displayAllData,
                displayAllData = _ref2$displayAllData === undefined ? this.props.displayAllData : _ref2$displayAllData;

            var numberRows = rowsPerPage || this.props.rowsPerPage;
            var toPage = page || this.state.currentPage,
                indexStart = displayAllData ? 0 : (toPage - 1) * numberRows,
                dataToSlice = data || this.props.data;
            if (displayAllData) {
                numberRows = numberRows * toPage;
            }
            if (orderBy) {
                dataToSlice = _.orderBy(dataToSlice, orderBy);
            }
            this.tableData = dataToSlice.slice(indexStart, indexStart + numberRows);
            // console.log('set table data', this.props.data, this.tableData);
        }
    }, {
        key: 'setPage',
        value: function setPage(page) {
            if (page === 0 || page > this.pages.length) return;
            this.setState({ currentPage: page });
        }
    }, {
        key: 'loadDataByHeight',
        value: function loadDataByHeight() {
            if (!this.props.displayAllData) return;
            var bodyTable = this.tableBodyComponent.table;
            var bodyRow = this.tableBodyComponent.table.childNodes[0];
            var tableBodyContainer = this.tableBodyComponent.table;
            var rowHeigh = bodyRow.offsetHeight;
            var tableBodyContainerHeight = tableBodyContainer.offsetHeight;

            if (!bodyRow || !rowHeigh || bodyTable.offsetHeight > tableBodyContainer.offsetHeight) return;

            this.setPage(Math.ceil(Math.ceil(tableBodyContainerHeight / rowHeigh) / this.props.rowsPerPage));
        }
    }, {
        key: 'render',
        value: function render() {
            var _this2 = this;

            // console.log('DataTable rendered', this.tableData.length, this.props.data.length);
            var pagesCount = this.getPagesCount();
            return _react2.default.createElement(
                'div',
                { className: (0, _classnames2.default)('data-table', this.props.className, { 'fixed-header': this.props.fixedHeader }), onClick: this.props.onClick, ref: function ref(_ref3) {
                        return _this2.container = _ref3;
                    } },
                _react2.default.createElement(
                    'div',
                    { className: 'table-container' },
                    _react2.default.createElement(
                        'div',
                        { className: (0, _classnames2.default)({ table: !this.props.fixedHeader }) },
                        _react2.default.createElement(_TableHeader2.default, {
                            columns: this.props.columns,
                            orderBy: this.state.orderBy,
                            onAction: this.onHeaderAction,
                            fixedHeader: this.props.fixedHeader,
                            ref: function ref(node) {
                                return _this2.tableHeaderComponent = node;
                            } }),
                        _react2.default.createElement(_TableBody2.default, {
                            data: this.tableData,
                            fixedHeader: this.props.fixedHeader,
                            columns: this.props.columns,
                            onRowRender: this.props.onRowRender,
                            scrollHandler: this.bodyScrollHandler,
                            ref: function ref(node) {
                                return _this2.tableBodyComponent = node;
                            } })
                    )
                ),
                this.props.displayPager && !this.props.displayAllData ? _react2.default.createElement(
                    'div',
                    { className: 'data-table-pager' },
                    _react2.default.createElement(
                        'button',
                        { className: (0, _classnames2.default)('previous', { hidden: this.state.currentPage === 1 }), onClick: this.previousClickHandler },
                        'Previous'
                    ),
                    _react2.default.createElement(
                        'div',
                        { className: 'pager' },
                        _react2.default.createElement(
                            'select',
                            { onChange: this.selectPageHandler, value: this.state.currentPage },
                            this.pages.map(function (pageNumber) {
                                return _react2.default.createElement(
                                    'option',
                                    { value: pageNumber, key: pageNumber },
                                    pageNumber
                                );
                            })
                        ),
                        _react2.default.createElement(
                            'span',
                            null,
                            ' / ',
                            pagesCount
                        )
                    ),
                    _react2.default.createElement(
                        'button',
                        { className: (0, _classnames2.default)('next', { hidden: pagesCount === 1 || pagesCount - this.state.currentPage === 0 }), onClick: this.nextClickHandler },
                        'Next'
                    )
                ) : null
            );
        }
    }]);
    return DataTable;
}(_react.Component);
// import type {ColumnT, DataT} from './Types';

// $FlowSkipCheck


DataTable.defaultProps = {
    rowsPerPage: 5,
    displayAllData: false,
    displayPager: true,
    firstIndex: 0
};
exports.default = DataTable;