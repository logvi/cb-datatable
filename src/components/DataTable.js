import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import * as _ from '../utils';

import TableBody from './TableBody';
import TableHeader from './TableHeader';
import Pager from './Pager';

class DataTable extends Component {
    static propTypes = {
        data: PropTypes.array,
        columns: PropTypes.array,
        rowsPerPage: PropTypes.number,
        orderBy: PropTypes.object,
        className: PropTypes.string,
        fixedHeader: PropTypes.bool,
        displayAllData: PropTypes.bool,
        onClick: PropTypes.func,
        onRowRender: PropTypes.func
    };

    static defaultProps = {
        rowsPerPage: 5,
		displayAllData: false,
		displayPager: true,
        firstIndex: 0
    };

    constructor(props) {
        super(props);
        this.state = {
            currentPage: 1,
            orderBy: props.orderBy,
            firstIndex: props.firstIndex
        };

        this.setTableData({
            page: this.state.currentPage,
            orderBy: this.state.orderBy
        });
		this.initPagesArray();
	}

	componentDidMount() {
	    if (this.props.fixedHeader) this.adjustColumnsSize();

        this.container.addEventListener('scroll', this.containerScrollHandler);
    }

    containerScrollHandler = (e) => {
        console.log(e);
    };

    componentWillUpdate(nextProps, nextState) {
    	// console.time('datatable');
        let pageChanged = this.state.currentPage !== nextState.currentPage,
            sortingChanged = this.state.orderBy !== nextState.orderBy;

        if (pageChanged || sortingChanged) {
            this.setTableData({
            	data: nextProps.data,
                page:  pageChanged ? nextState.currentPage : null,
                orderBy: sortingChanged ? nextState.orderBy : this.state.orderBy
            });
			return;
        }

        this.setTableData({data: nextProps.data, orderBy: this.state.orderBy});
		this.initPagesArray(nextProps.data);
	}

	componentDidUpdate() {
        if (this.props.fixedHeader) {
            // debugger;
            this.adjustColumnsSize();
        }
    }

	getPagesCount(data) {
		let _data = data || this.props.data;
        return Math.ceil(_data.length / this.props.rowsPerPage);
    }

    initPagesArray(data) {
    	let _data =  data || this.props.data;
        this.pages = Array.from(new Array(this.getPagesCount(_data)), (val, index) => index + 1);
    }

    setTableData({data, page, orderBy, rowsPerPage, displayAllData = this.props.displayAllData}) {
    	let numberRows = rowsPerPage || this.props.rowsPerPage;
        let toPage = page || this.state.currentPage,
            indexStart = displayAllData ? 0 : (toPage - 1) * numberRows,
            dataToSlice = data || this.props.data;
		if (displayAllData) {
			numberRows = numberRows * toPage;
		}
        if (orderBy) {
            dataToSlice = _.orderBy(dataToSlice, orderBy);
        }
        this.tableData = dataToSlice.slice(indexStart, indexStart + numberRows);
		// console.log('set table data', this.props.data, this.tableData);
    }

    setPage(page) {
        if (page === 0 || page > this.pages.length) return;
        this.setState({currentPage: page});
    }

	selectPageHandler = (e) => {
		e.preventDefault();
		let toPage = e.target.value;
		this.setState({currentPage: +toPage});
	};

    onHeaderAction = (type, params) => {
        switch (type) {
            case 'sort':
                const newSorting = {[params.column.name]: params.orderBy};
                this.setState({orderBy: newSorting});
                break;
            default:
                break;
        }
    };

	bodyScrollHandler = ({target}) => {
		const {scrollLeft} = target;
		if (this.props.fixedHeader) {
			this.tableHeaderComponent.table.style.transform = 'translate3d(' + (-1 * scrollLeft) + 'px, 0, 0)';
		}
		if (this.props.displayAllData) {
			const {scrollTop, offsetHeight, scrollHeight} = target;
			const scrollPosToUpdate = scrollHeight * 0.15;

			if (scrollHeight - (scrollTop + offsetHeight) <= scrollPosToUpdate) {
				this.setPage(this.state.currentPage + 1);
			}
		}
	};

    adjustColumnsSize = () => {
        if (!this.props.fixedHeader) return;
        // this.tableHeaderComponent.table.style.width = this.tableHeaderComponent.fixedContainer.offsetWidth + 'px';
        let headerColumns = this.tableHeaderComponent.row.childNodes;
        // debugger;
        for (let j=0; j < headerColumns.length; j++) {
            let bodyColumn = this.tableBodyComponent.table.childNodes[0].childNodes[j],
                headerColumn = headerColumns[j];
            // clear minWidth
            if (!!bodyColumn.style.width) {
                bodyColumn.style.width = '';
                bodyColumn.style.minWidth = '';
                headerColumn.style.width = '';
            }
            var bodyColumnWidth = bodyColumn.getBoundingClientRect().width,
                headerColumnWidth = headerColumn.getBoundingClientRect().width,
                newWidth = bodyColumnWidth;
            // newWidth = Math.max(bodyColumnWidth, headerColumnWidth);
            headerColumn.style.width = newWidth + 'px';
            if (headerColumn.getBoundingClientRect().width > newWidth) {
                newWidth = Math.max(bodyColumnWidth, headerColumnWidth);
                headerColumn.style.width = newWidth + 'px';
            }
            bodyColumn.style.minWidth = newWidth + 'px';
            bodyColumn.style.width = newWidth + 'px';
        }
        // this.tableHeaderComponent.table.style.width = 'auto';
    };

    previousClickHandler = () => {
        this.setPage(this.state.currentPage - 1);
    };

    nextClickHandler = () => {
        this.setPage(this.state.currentPage + 1);
    };

	loadDataByHeight() {
		if (!this.props.displayAllData) return;
		const bodyTable = this.tableBodyComponent.table;
		const bodyRow = this.tableBodyComponent.table.childNodes[0];
		const tableBodyContainer = this.tableBodyComponent.table;
		const rowHeigh = bodyRow.offsetHeight;
		const tableBodyContainerHeight = tableBodyContainer.offsetHeight;

		if (!bodyRow || !rowHeigh || bodyTable.offsetHeight > tableBodyContainer.offsetHeight) return;

		this.setPage(Math.ceil(Math.ceil(tableBodyContainerHeight/rowHeigh) / this.props.rowsPerPage));
	}

    render() {
        // console.log('DataTable rendered', this.tableData.length, this.props.data.length);
		let pagesCount = this.getPagesCount();
        return (
            <div className={classNames('data-table', this.props.className, {'fixed-header': this.props.fixedHeader})} onClick={this.props.onClick} ref={ref => this.container = ref}>
                <div className="table-container">
                    <div className={classNames({table: !this.props.fixedHeader})}>
                        <TableHeader
                            columns={this.props.columns}
                            orderBy={this.state.orderBy}
                            onAction={this.onHeaderAction}
                            fixedHeader={this.props.fixedHeader}
                            ref={node => this.tableHeaderComponent = node}/>

                        <TableBody
                            data={this.tableData}
                            fixedHeader={this.props.fixedHeader}
                            columns={this.props.columns}
                            onRowRender={this.props.onRowRender}
                            scrollHandler={this.bodyScrollHandler}
                            ref={node => this.tableBodyComponent = node} />
                    </div>
                </div>
                {this.props.displayPager && !this.props.displayAllData ?
					<div className="data-table-pager">
						<button className={classNames('previous', {hidden: this.state.currentPage === 1})} onClick={this.previousClickHandler}>Previous</button>
						<div className="pager">
							<select onChange={this.selectPageHandler} value={this.state.currentPage}>
								{this.pages.map(pageNumber => <option value={pageNumber} key={pageNumber}>{pageNumber}</option>)}
							</select>
							<span> / {pagesCount}</span>
						</div>
						<button className={classNames('next', {hidden: pagesCount === 1 || pagesCount - this.state.currentPage === 0})} onClick={this.nextClickHandler}>Next</button>
					</div>
				: null }
            </div>
        );
    }
}

export default DataTable;