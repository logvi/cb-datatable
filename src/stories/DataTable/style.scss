.content {
  position: relative;

  .datatable {
    overflow: auto;
  }

  .data-table {
    width: 100%;

    .data-table-header {
      .column-header {
        border: 1px solid #131217;
        text-align: center;
        cursor: pointer;
        color: rgba(0, 0, 0, .54);
        background-image: linear-gradient(to bottom, #fafafa 2%, #efefef 98%);
      }
    }

    .table-cell {
      border: 1px solid #131217;
      white-space: nowrap;
      padding: 5px;
    }
  }
}