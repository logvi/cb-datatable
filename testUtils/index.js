function generateMockData({columnsNumber = 4, columnLabel = 'column', rowsNumber = 100, dataFunc = false}) {
  let columns = [];
  for (let i = 0; i < columnsNumber; i++) {
    let columnName =  'column' + i,
        _columnLabel = typeof columnLabel === 'function' ? columnLabel(i) : columnLabel + ' ' + i;
    columns.push({
      name: columnName,
      label: _columnLabel,
      className: columnName
    })
  }

  let data = [];
  for (let i = 0; i < rowsNumber; i++) {
    let row = {};
    columns.forEach(column => {
      row[column.name] = dataFunc ? dataFunc(column, i) : column.name + ' ' + i;
    });
    data.push(row);
  }
  return {columns, data};
}

function generateDataForColumns(columns, rowsNumber, dataFunc = false) {
  let data = [];
  for (let i = 0; i < rowsNumber; i++) {
    let row = {};
    columns.forEach(column => {
      row[column.name] = dataFunc ? dataFunc(column, i) : column.name + ' ' + i;
    });
    data.push(row);
  }
  return data;
}

export {
  generateMockData,
  generateDataForColumns
};
