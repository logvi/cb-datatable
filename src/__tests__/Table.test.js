import React from 'react';
import {generateMockData} from '../../testUtils';
import sinon from 'sinon';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import DataTable from '../';

describe('DataTable', () => {
  const {columns, data} = generateMockData({columnsNumber: 10, rowsNumber: 50});

  it('Wrapper div should have className from props', () => {
    const tree = shallow(
      <DataTable columns={columns} data={data} className="test" />
    );
    expect(tree.hasClass('test')).toBeTruthy();
  });

  it('onClick prop should fire', () => {
    const clickHandler = sinon.spy();

    const tree = shallow(
      <DataTable columns={columns} data={data} onClick={clickHandler} />
    );
    tree.simulate('click');
    expect(clickHandler.called).toBeTruthy();
  });

  it('should apply renderLayout prop, it displays only table', () => {
    const {columns, data} = generateMockData({columnsNumber: 1, rowsNumber: 1});
    const tableLayout = (Table, Pager) => {
      return Table;
    };
    const tree = renderer.create(
      <DataTable columns={columns} data={data} renderLayout={tableLayout} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('orderBy asc', () => {
    const {columns, data} = generateMockData({columnsNumber: 1, rowsNumber: 3});
    const tree = shallow(<DataTable columns={columns} data={data} orderBy={{column0: 'asc'}} />);
    const instance = tree.instance();
    expect(instance.data).toEqual([
      {column0: 'column0 0'},
      {column0: 'column0 1'},
      {column0: 'column0 2'},
    ]);
  });

  it('orderBy asc Snapshot: should apply sort class and icon', () => {
    const {columns, data} = generateMockData({columnsNumber: 1, rowsNumber: 3});
    const tree = renderer.create(<DataTable columns={columns} data={data} orderBy={{column0: 'asc'}} PagerComponent={null}/>);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('orderBy desc', () => {
    const {columns, data} = generateMockData({columnsNumber: 1, rowsNumber: 3});
    const tree = shallow(<DataTable columns={columns} data={data} orderBy={{column0: 'desc'}} />);
    const instance = tree.instance();
    expect(instance.data).toEqual([
      {column0: 'column0 2'},
      {column0: 'column0 1'},
      {column0: 'column0 0'},
    ]);
  });

  it('orderBy desc Snapshot: should apply sort class and icon', () => {
    const {columns, data} = generateMockData({columnsNumber: 1, rowsNumber: 3});
    const tree = renderer.create(<DataTable columns={columns} data={data} orderBy={{column0: 'desc'}} PagerComponent={null}/>);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('Should update data when data from props is changed, orderBy should be applied', () => {
    let {columns, data} = generateMockData({columnsNumber: 1, rowsNumber: 3});
    const tree = shallow(<DataTable columns={columns} data={data} orderBy={{column0: 'desc'}} PagerComponent={null}/>);
    const instance = tree.instance();
    const newData = generateMockData({columnsNumber: 1, rowsNumber: 4});
    data.push(newData.data[newData.data.length - 1]);
    tree.setProps({data: data});
    expect(instance.data[0]).toEqual({
      column0: 'column0 3'
    });
  });
});
