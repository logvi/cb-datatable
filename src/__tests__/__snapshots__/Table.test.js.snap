// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Columns visibility. Column 1 should be hidden 1`] = `
<div
  className="cb-datatable"
>
  <div
    className="cb-datatable__table"
  >
    <table>
      <thead>
        <tr>
          <th
            className="sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column0
            </div>
          </th>
          <th
            className="sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column2
            </div>
          </th>
          <th
            className="sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column3
            </div>
          </th>
          <th
            className="sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column4
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          onClick={null}
        >
          <td
            className={null}
          >
            <div
              className="table-cell-content"
            >
              column0 0
            </div>
          </td>
          <td
            className={null}
          >
            <div
              className="table-cell-content"
            >
              column2 0
            </div>
          </td>
          <td
            className={null}
          >
            <div
              className="table-cell-content"
            >
              column3 0
            </div>
          </td>
          <td
            className={null}
          >
            <div
              className="table-cell-content"
            >
              column4 0
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`DataTable orderBy asc Snapshot: should apply sort class and icon 1`] = `
<div
  className="cb-datatable"
>
  <div
    className="cb-datatable__table"
  >
    <table>
      <thead>
        <tr>
          <th
            className="column0 sortable sort-asc"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              <div
                className="th-content__label"
              >
                column 0
              </div>
              <span
                className="icon-sort icon-sort-asc"
              >
                 ▲
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 0
            </div>
          </td>
        </tr>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 1
            </div>
          </td>
        </tr>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 2
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`DataTable orderBy desc Snapshot: should apply sort class and icon 1`] = `
<div
  className="cb-datatable"
>
  <div
    className="cb-datatable__table"
  >
    <table>
      <thead>
        <tr>
          <th
            className="column0 sortable sort-desc"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              <div
                className="th-content__label"
              >
                column 0
              </div>
              <span
                className="icon-sort icon-sort-desc"
              >
                 ▼
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 2
            </div>
          </td>
        </tr>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 1
            </div>
          </td>
        </tr>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 0
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`DataTable should apply renderLayout prop, it displays only table 1`] = `
<div
  className="cb-datatable"
>
  <div
    className="cb-datatable__table"
  >
    <table>
      <thead>
        <tr>
          <th
            className="column0 sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column 0
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          onClick={null}
        >
          <td
            className="column0"
          >
            <div
              className="table-cell-content"
            >
              column0 0
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
